### PR TITLE
Performance optimizations for multiclass adaboost

### PR DESCRIPTION
**Pull request by orazve**
What has been done:
- Disabled checking for algorithms which are called from kernel
- Added parallelism into prediction algorithm
- Optimized memory usage for prediction stage
- Optimized weighted gini computation
- CSR labels table works correctly
PCM: http://intel-ci.intel.com/ea0cfcb4-5320-f198-a1a1-8cdcd4b723a1

### CODE REVIEW

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 1]
```diff
@@ -196,41 +243,78 @@ services::Status AdaBoostPredictKernel<method, algorithmFPType, cpu>::computeCom
...
     for(size_t k = 0; k < nClasses; k++)
     {
-        for(size_t m = 0; m < nWeakLearners; m++)
+        DAAL_CHECK_STATUS(s, computeClassScore(k, nVectors, nClasses, weakPredictions, curClassScore, maxClassScore, r, alpha, nWeakLearners));
+    }
+    if(nClasses == 2)
+    {
+        const algorithmFPType minusOne = (algorithmFPType) - 1.0;
+        const algorithmFPType zero = (algorithmFPType)0.0;
+        for(size_t j = 0; j < nVectors; j++)
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/adaboost/adaboost_predict_impl.i]: It's seq loop. Does it affect performance of inference?

File: algorithms/kernel/adaboost/adaboost_predict_impl.i [hunk 2]
```diff
@@ -94,35 +96,79 @@ services::Status AdaBoostPredictKernel<method, algorithmFPType, cpu>::computeImp
...
-    algorithmFPType *pSumLog = pSumLogArray.get();
-    service_memset<algorithmFPType, cpu>(pSumLog, 0, nVectors);
 
+    size_t nBlocks = nVectors / _nRowsInBlock;
+    nBlocks += (nBlocks * _nRowsInBlock != nVectors);
+
+    SafeStatus safeStat;
+    daal::threader_for(nBlocks, nBlocks, [ =, &safeStat, &threadBuffer](int block)
+    {
+        size_t nRowsToProcess = _nRowsInBlock;
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/adaboost/adaboost_predict_impl.i]: Can be a bit shorter:
```cpp 
const size_t nRowsToProcess = ((block == nBlocks - 1) ? (nVectors - block * _nRowsInBlock) : _nRowsInBlock);
```

File: algorithms/kernel/adaboost/adaboost_train_impl.i [hunk 3]
```diff
@@ -98,6 +98,7 @@ services::Status AdaBoostTrainKernel<method, algorithmFPType, cpu>::adaboostSAMM
     service_memset<algorithmFPType, cpu>(w, invNVectors, nVectors);
 
     services::SharedPtr<classifier::training::Batch> learnerTrain = parameter->weakLearnerTraining->clone();
+    learnerTrain->enableChecks(false);
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/adaboost/adaboost_train_impl.i]: We checked this before and don't want to check this again, right?

File: algorithms/kernel/stump/stump_classification_predict_impl.i [hunk 4]
```diff
@@ -53,13 +53,16 @@ template <Method method, typename algorithmFPtype, CpuType cpu>
...
-    const size_t nVectors  = yTable->getNumberOfRows();
-    WriteColumns<algorithmFPtype, cpu> y(const_cast<NumericTable *>(yTable), 0, 0, nVectors);
+    const size_t nVectors = yTable->getNumberOfRows();
+    WriteColumns<algorithmFPtype, cpu> y(yTable, 0, 0, nVectors);
     DAAL_CHECK_STATUS(s, y.status());
     algorithmFPtype *yArray = y.get();
+    DAAL_CHECK(yArray, ErrorMemoryAllocationFailed);
+    const algorithmFPtype zero = 0.0;
+    const algorithmFPtype minusOne = -1.0;
     for(size_t i = 0; i < nVectors; i++)
```
Comment (SmirnovEgorRu) [inline algorithms/kernel/stump/stump_classification_predict_impl.i]: Hotspot? Threading can be added

File: algorithms/kernel/decision_tree/decision_tree_classification_split_criterion.i [hunk 5]
```diff
@@ -448,31 +448,31 @@ struct GiniWeighted
         const ValueType invLeftWeight = static_cast<ValueType>(1) / leftWeight;
         const ValueType leftProbability = leftWeight * invTotalWeight;
         const ValueType rightProbability = rightWeight * invTotalWeight;
-        ValueType leftGini = 1;
-        ValueType rightGini = 1;
+        ValueType leftGini = leftWeight * leftWeight;
```
Comment (ShvetsKS) [inline algorithms/kernel/decision_tree/decision_tree_classification_split_criterion.i]: it`s not obvious why initial value was changed.
Could you add comment please?

File: algorithms/kernel/adaboost/adaboost_predict_kernel.h [hunk 6]
```diff
@@ -43,16 +44,36 @@ namespace prediction
...
+
+    services::Status processBlockClassScore(
+        size_t nProcessedRows,
+        size_t nRowsInCurrentBlock,
+        const size_t k, const size_t nClasses,
+        daal::services::Collection<services::SharedPtr<daal::internal::HomogenNumericTableCPU<algorithmFPType, cpu> > > &weakPredictions,
+        algorithmFPType *curClassScore, algorithmFPType *maxClassScore, algorithmFPType *r, const algorithmFPType *alpha, const size_t nWeakLearners);
+
+protected:
+    static const size_t _nRowsInBlock = 5000;
```
Comment (ShvetsKS) [inline algorithms/kernel/adaboost/adaboost_predict_kernel.h]: can we reuse daal::algorithms::dtrees::prediction::internal::TileDimensions struct to have flexibility related to number of features in dataset?

File: algorithms/kernel/stump/stump_classification_train_aux.i [hunk 7]
```diff
@@ -51,79 +52,82 @@ namespace internal
...
-    WriteColumns<algorithmFPtype, cpu> y(const_cast<NumericTable *>(yTable), 0, 0, nVectors);
-    DAAL_CHECK_STATUS(s, y.status());
-    algorithmFPtype *yArray = y.get();
+    const algorithmFPtype one = 1.0;
+    const algorithmFPtype zero = 0.0;
+    const algorithmFPtype minusOne = -1.0;
     for(size_t i = 0; i < nVectors; i++)
     {
-        if(yArray[i] == -1) { yArray[i] = 0; }
+        if(yArray[i] == minusOne) { yZeroOne[i] = zero; }
```
Comment (averbukh) [inline algorithms/kernel/stump/stump_classification_train_aux.i]: Add comment about usage of == for fp type

File: algorithms/kernel/adaboost/adaboost_predict_impl.i (hunk 2 above)
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: thanks, will be changed

File: algorithms/kernel/adaboost/adaboost_train_impl.i (hunk 3 above)
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_train_impl.i]: yes

File: algorithms/kernel/decision_tree/decision_tree_classification_split_criterion.i (hunk 5 above)
Comment (orazve) [reply algorithms/kernel/decision_tree/decision_tree_classification_split_criterion.i]: @ShvetsKS 
gini index can be rewritten in this way:
image
where W - sum of weights of all classes, w_k - weight of k-th class, C - number of classes

File: algorithms/kernel/stump/stump_classification_predict_impl.i (hunk 4 above)
Comment (orazve) [reply algorithms/kernel/stump/stump_classification_predict_impl.i]: It is not hotspot, just removing extra conversions.

File: algorithms/kernel/stump/stump_classification_train_aux.i (hunk 7 above)
Comment (orazve) [reply algorithms/kernel/stump/stump_classification_train_aux.i]: ok

File: algorithms/kernel/adaboost/adaboost_predict_impl.i (hunk 2 above)
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_predict_impl.i]: thanks, it is done

File: algorithms/kernel/adaboost/adaboost_train_impl.i (hunk 3 above)
Comment (orazve) [reply algorithms/kernel/adaboost/adaboost_train_impl.i]: yes

**Comment by orazve** [comment]
New CI: http://intel-ci.intel.com/ea172589-8ded-f1ac-bf41-8cdcd4b723a1

**Comment by ShvetsKS** [comment]
make test

**Comment by orazve** [comment]
make test

**Pull request merged**